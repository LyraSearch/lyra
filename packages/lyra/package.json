{
  "name": "@nearform/lyra",
  "description": "Fast, in-memory, typo-tolerant, full-text search engine",
  "version": "0.0.1-beta-17",
  "author": {
    "name": "Michele Riva",
    "email": "ciao@micheleriva.it",
    "url": "https://github.com/MicheleRiva"
  },
  "bugs": {
    "url": "https://github.com/nearform/lyra"
  },
  "license": "Apache-2.0",
  "scripts": {
    "build:cjs": "tsc --project tsconfig.cjs.json",
    "build:module": "tsc --project tsconfig.esm.json",
    "build": "rimraf ./dist && pnpm build:module && pnpm build:cjs",
    "test": "c8 -c tests/config/c8-local.json tap --rcfile=tests/config/tap.yml tests/*.test.ts",
    "test:ci": "c8 -c tests/config/c8-ci.json tap --rcfile=tests/config/tap.yml --no-color tests/*.test.ts",
    "ci": "npm run build && npm run test:ci"
  },
  "main": "./dist/cjs/lyra.js",
  "module": "./dist/esm/lyra.js",
  "types": "./dist/esm/lyra.d.ts",
  "dependencies": {
    "fastq": "^1.13.0"
  },
  "devDependencies": {
    "@types/node": "^18.6.2",
    "@types/tap": "^15.0.7",
    "c8": "^7.12.0",
    "rimraf": "^3.0.2",
    "tap": "^16.3.0",
    "ts-node": "^10.9.1",
    "typescript": "^4.7.4"
  }
}