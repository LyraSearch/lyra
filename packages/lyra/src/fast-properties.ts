/*
 * Adapred from: https://github.com/sindresorhus/to-fast-properties
 *
 * Original license:
 *
 * MIT License
 *
 * Copyright (c) Petka Antonov
 * Copyright (c) Benjamin Gruenbaum
 * Copyright (c) John-David Dalton
 * Copyright (c) Sindre Sorhus
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let fastProto: any = null;

function FastObject<T extends object>(object?: T): T {
  if (fastProto !== null && typeof fastProto.property) {
    const result = fastProto;
    fastProto = FastObject.prototype = null;
    return result;
  }

  fastProto = FastObject.prototype = object == null ? Object.create(null) : object;

  // @ts-expect-error this is a hack to make the type checker happy (written by Copilot LOL)
  return new FastObject();
}

const inlineCacheCutoff = 10;

for (let index = 0; index <= inlineCacheCutoff; index++) {
  FastObject();
}

export function insertWithFastProperties<T extends object>(object: T, propName: keyof T, value: T[keyof T]): T {
  object[propName] = value;
  toFastProperties(object);
  return object;
}

export default function toFastProperties<T extends object>(object?: T): T {
  return FastObject(object);
}
