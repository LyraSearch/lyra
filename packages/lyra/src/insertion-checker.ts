const kInsertions = Symbol("lyra.insertions");

const warn =
  "process" in globalThis
    ? process.emitWarning
    : function emitWarning(message: string, options: { code: string }) {
        console.warn(`[WARNING] [${options.code}] ${message}`);
      };

export function trackInsertion(_lyra: unknown) {
  const lyra = _lyra as object & { [kInsertions]?: number };

  if (typeof lyra[kInsertions] !== "number") {
    queueMicrotask(() => {
      lyra[kInsertions] = undefined;
    });

    lyra[kInsertions] = 0;
  }

  if (lyra[kInsertions]! > 1000) {
    warn(
      "Lyra's insert operation is synchronous. Please avoid inserting a large number of document in a single operation in order not to block the main thread.",
      { code: "LYRA0001" },
    );

    lyra[kInsertions] = -1;
  } else if (lyra[kInsertions] >= 0) {
    lyra[kInsertions]++;
  }
}
