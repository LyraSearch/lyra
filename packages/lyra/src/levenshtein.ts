export function levenshtein(a: string, b: string): number {
  /* c8 ignore next 3 */
  if (!a.length) {
    return b.length;
  }

  /* c8 ignore next 3 */
  if (!b.length) {
    return a.length;
  }

  let tmp;

  if (a.length > b.length) {
    tmp = a;
    a = b;
    b = tmp;
  }

  const row = Array.from({ length: a.length + 1 }, (_, i) => i);
  let val = 0;

  for (let i = 1; i <= b.length; i++) {
    let prev = i;

    for (let j = 1; j <= a.length; j++) {
      if (b[i - 1] === a[j - 1]) {
        val = row[j - 1];
      } else {
        val = Math.min(row[j - 1] + 1, Math.min(prev + 1, row[j] + 1));
      }

      row[j - 1] = prev;
      prev = val;
    }
    row[a.length] = prev;
  }

  return row[a.length];
}
