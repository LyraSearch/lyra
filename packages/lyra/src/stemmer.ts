import {
  PorterStemmer,
  PorterStemmerNl,
  PorterStemmerFr,
  PorterStemmerEs,
  PorterStemmerIt,
  PorterStemmerPt,
  PorterStemmerSv,
  PorterStemmerRu,
  PorterStemmerNo,
  Stemmer,
} from "natural";

export type Language = typeof SUPPORTED_LANGUAGES[number];

export const SUPPORTED_LANGUAGES = [
  "dutch",
  "english",
  "french",
  "italian",
  "norwegian",
  "portugese",
  "russian",
  "spanish",
  "swedish",
] as const;

export function stemArray(input: string[], language: Language): string[] {
  let stemmer: Stemmer;

  switch (language) {
    case "dutch":
      stemmer = PorterStemmerNl;
      break;
    case "english":
      stemmer = PorterStemmer;
      break;
    case "french":
      stemmer = PorterStemmerFr;
      break;
    case "italian":
      stemmer = PorterStemmerIt;
      break;
    case "norwegian":
      stemmer = PorterStemmerNo;
      break;
    case "portugese":
      stemmer = PorterStemmerPt;
      break;
    case "russian":
      stemmer = PorterStemmerRu;
      break;
    case "spanish":
      stemmer = PorterStemmerEs;
      break;
    case "swedish":
      stemmer = PorterStemmerSv;
      break;
    default:
      stemmer = PorterStemmer;
      break;
  }

  return input.map(word => stemmer.stem(word));
}
