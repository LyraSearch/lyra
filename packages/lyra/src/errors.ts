import { SUPPORTED_LANGUAGES } from "./stemmer";

function formatJSON(input: object) {
  return JSON.stringify(input, null, 2);
}

export function INVALID_SCHEMA_TYPE(type: string): string {
  return `Invalid schema type. Expected string or object, but got ${type}`;
}

export function INVALID_DOC_SCHEMA(expected: object, found: object): string {
  return `Invalid document structure. \nClass has been initialized with the following schema: \n\n${formatJSON(
    expected,
  )}\n\nbut found the following doc:\n\n${formatJSON(found)}`;
}

export function INVALID_PROPERTY(name: string, expected: string[]): string {
  return `Invalid property name. Expected a wildcard string ("*") or array containing one of the following properties: ${expected.join(
    ", ",
  )}, but got: ${name}`;
}

export function CANT_DELETE_DOC_NOT_FOUND(id: string): string {
  return `Document with ID ${id} does not exist.`;
}

export function CANT_DELETE_DOCUMENT(docID: string, key: string, token: string): string {
  return `Unable to delete document "${docID}" from index "${key}" on word "${token}".`;
}

export function UNSUPPORTED_NESTED_PROPERTIES(): string {
  return `Nested properties are not supported in this Lyra version, but will be in the future.`;
}

export function DOC_ID_DOES_NOT_EXISTS(id: string): string {
  return `Document with ID ${id} does not exists`;
}

export function LANGUAGE_NOT_SUPPORTED(lang: string): string {
  return `Language "${lang}" is not supported.\nSupported languages are:\n - ${SUPPORTED_LANGUAGES.join("\n - ")}`;
}

export function GETTER_SETTER_WORKS_ON_EDGE_ONLY(method: string): string {
  return `${method} works on edge only. Use edge: true in Lyra constructor to enable it.`;
}
