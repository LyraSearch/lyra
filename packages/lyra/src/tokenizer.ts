import { WordTokenizer } from "natural";
import { Language, stemArray } from "./stemmer";

const tokenizer = new WordTokenizer();

export function tokenize(
  input: string,
  language: Language = "english",
  enableStemming = true
): Set<string> {
  const tokens = tokenizer.tokenize(input);

  if (!enableStemming) {
    return new Set(tokens);
  }

  return new Set(stemArray(tokens, language));
}
