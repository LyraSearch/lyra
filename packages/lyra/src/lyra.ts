import type { queueAsPromised } from "fastq";
import { nanoid } from "nanoid";
import * as fastq from "fastq";
import { Trie } from "./prefix-tree/trie";
import * as ERRORS from "./errors";
import { tokenize } from "./tokenizer";
import { formatNanoseconds, getNanosecondsTime } from "./utils";
import { Language, SUPPORTED_LANGUAGES } from "./stemmer";

export type PropertyType = "string" | "number" | "boolean";

export type PropertiesSchema = {
  [key: string]: PropertyType | PropertiesSchema;
};

export type LyraProperties = {
  schema: PropertiesSchema;
  defaultLanguage?: Language;
};

export type LyraDocs = Map<string, object>;

export type SearchParams = {
  term: string;
  properties?: "*" | string[];
  limit?: number;
  offset?: number;
  exact?: boolean;
};

type LyraIndex = Map<string, Trie>;

type QueueDocParams = {
  id: string;
  doc: object;
  language: Language;
};

type SearchResult = Promise<{
  count: number;
  hits: object[];
  elapsed: string;
}>;

export class Lyra {
  private defaultLanguage: Language = "english";
  private schema: PropertiesSchema;
  private docs: LyraDocs = new Map();
  private index: LyraIndex = new Map();
  private queue: queueAsPromised<QueueDocParams> = fastq.promise(
    this,
    this._insert,
    1
  );

  constructor(properties: LyraProperties) {
    const defaultLanguage =
      (properties?.defaultLanguage?.toLowerCase() as Language) ?? "english";

    if (!SUPPORTED_LANGUAGES.includes(defaultLanguage)) {
      throw ERRORS.LANGUAGE_NOT_SUPPORTED(defaultLanguage);
    }

    this.defaultLanguage = defaultLanguage;
    this.schema = properties.schema;
    this.buildIndex(properties.schema);
  }

  private buildIndex(schema: PropertiesSchema) {
    for (const prop in schema) {
      const propType = typeof prop;

      if (propType === "string") {
        this.index.set(prop, new Trie());
      } else if (propType === "object") {
        // @todo nested properties will be supported with the next versions
        //this.buildIndex(prop as unknown as PropertiesSchema);
        throw ERRORS.UNSUPPORTED_NESTED_PROPERTIES();
      } else {
        throw ERRORS.INVALID_SCHEMA_TYPE(propType);
      }
    }
  }

  async search(
    params: SearchParams,
    language: Language = this.defaultLanguage
  ): SearchResult {
    const tokens = tokenize(params.term, language).values();
    const indices = this.getIndices(params.properties);
    const { limit = 10, offset = 0, exact = false } = params;
    const results: object[] = new Array({ length: limit });
    let totalResults = 0;

    const timeStart = getNanosecondsTime();

    let i = 0;
    let j = 0;

    for (const token of tokens) {
      for (const index of indices) {
        const documentIDs = await this.getDocumentIDsFromSearch({
          ...params,
          index: index,
          term: token,
          exact: exact,
        });

        totalResults += documentIDs.size;

        if (i >= limit) {
          break;
        }

        if (documentIDs.size) {
          for (const id of documentIDs) {
            if (j < offset) {
              j++;
              continue;
            }

            if (i >= limit) {
              break;
            }

            const fullDoc = this.docs.get(id);
            results[i] = { id, ...fullDoc };
            i++;
          }
        }
      }
    }

    return {
      elapsed: formatNanoseconds(getNanosecondsTime() - timeStart),
      hits: results,
      count: totalResults,
    };
  }

  private getIndices(indices: SearchParams["properties"]): string[] {
    const knownIndices = [...this.index.keys()];

    if (!indices) {
      return knownIndices;
    }

    if (typeof indices === "string") {
      if (indices === "*") {
        return knownIndices;
      } else {
        throw ERRORS.INVALID_PROPERTY(indices, knownIndices);
      }
    }

    for (const index of indices as string[]) {
      if (!knownIndices.includes(index)) {
        throw ERRORS.INVALID_PROPERTY(index, knownIndices);
      }
    }

    return indices as string[];
  }

  async delete(docID: string): Promise<boolean> {
    if (!this.docs.has(docID)) {
      throw ERRORS.DOC_ID_DOES_NOT_EXISTS(docID);
    }

    const document = this.docs.get(docID)!;

    for (const key in document) {
      const idx = this.index.get(key)!;
      const tokens = tokenize((document as any)[key]);

      for (const token of tokens) {
        if (idx.removeDocByWord(token, docID)) {
          throw `Unable to remove document "${docID}" from index "${key}" on word "${token}".`;
        }
      }
    }

    this.docs.delete(docID);

    return true;
  }

  private async getDocumentIDsFromSearch(
    params: SearchParams & { index: string }
  ): Promise<Set<string>> {
    const idx = this.index.get(params.index);
    const searchResult = idx?.find({
      term: params.term,
      exact: params.exact,
    });
    const ids = new Set<string>();

    for (const key in searchResult) {
      for (const id of (searchResult as any)[key]) {
        ids.add(id);
      }
    }
    return ids;
  }

  public async insert(
    doc: object,
    language: Language = this.defaultLanguage
  ): Promise<{ id: string }> {
    const id = nanoid();

    if (!SUPPORTED_LANGUAGES.includes(language)) {
      throw ERRORS.LANGUAGE_NOT_SUPPORTED(language);
    }

    if (!(await this.checkInsertDocSchema(doc))) {
      throw ERRORS.INVALID_DOC_SCHEMA(this.schema, doc);
    }

    await this.queue.push({
      id,
      doc,
      language,
    });

    return { id };
  }

  private async _insert({ doc, id, language }: QueueDocParams): Promise<void> {
    const index = this.index;
    this.docs.set(id, doc);

    function recursiveTrieInsertion(doc: object) {
      for (const key in doc) {
        if (typeof (doc as any)[key] === "object") {
          throw ERRORS.UNSUPPORTED_NESTED_PROPERTIES();
          // @todo nested properties will be supported with the nexrt versions
          // recursiveTrieInsertion((doc as any)[key]);
        } else if (typeof (doc as any)[key] === "string") {
          const requestedTrie = index.get(key);
          const tokens = tokenize((doc as any)[key], language);

          for (const token of tokens) {
            requestedTrie?.insert(token, id);
          }
        }
      }
    }

    recursiveTrieInsertion(doc);
  }

  private async checkInsertDocSchema(
    doc: QueueDocParams["doc"]
  ): Promise<boolean> {
    function recursiveCheck(
      newDoc: QueueDocParams["doc"],
      schema: PropertiesSchema
    ): boolean {
      for (const key in newDoc) {
        if (!(key in schema)) {
          return false;
        }

        const propType = typeof (newDoc as any)[key];

        if (propType === "object") {
          recursiveCheck((newDoc as any)[key], schema);
        } else {
          if (typeof (newDoc as any)[key] !== schema[key]) {
            return false;
          }
        }
      }

      return true;
    }

    return recursiveCheck(doc, this.schema);
  }
}
