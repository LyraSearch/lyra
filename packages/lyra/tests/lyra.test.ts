import { Lyra } from "../src/lyra";

function getId({ id }: { id: string }) {
  return id;
}

describe("defaultLanguage", () => {
  it("should throw an error if the desired language is not supported", () => {
    try {
      new Lyra({
        schema: {},
        // @ts-expect-error latin is not supported
        defaultLanguage: "latin",
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it("should throw an error if the desired language is not supported during insertion", async () => {
    try {
      const db = new Lyra({
        schema: { foo: "string" },
      });

      await db.insert(
        {
          foo: "bar",
        },
        // @ts-expect-error latin is not supported
        "latin"
      );
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it("should not throw if if the language is supported", () => {
    try {
      new Lyra({
        schema: {},
        defaultLanguage: "portugese",
      });
    } catch (e) {
      expect(e).toBeUndefined();
    }
  });
});

describe("checkInsertDocSchema", () => {
  it("should compare the inserted doc with the schema definition", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
      },
    });

    expect(
      await db.insert({ quote: "hello, world!", author: "me" }).then(getId)
    ).toBeDefined();

    try {
      await db.insert({ quote: "hello, world!", author: true });
    } catch (err) {
      expect(err).toMatchSnapshot();
    }

    try {
      await db.insert({
        quote: "hello, world!",
        authors: "author should be singular",
      });
    } catch (err) {
      expect(err).toMatchSnapshot();
    }

    try {
      await db.insert({ quote: "hello, world!", foo: { bar: 10 } });
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });
});

describe("lyra", () => {
  it("should correctly insert and retrieve data", async () => {
    const db = new Lyra({
      schema: {
        example: "string",
      },
    });

    const ex1Insert = await db.insert({ example: "The quick, brown, fox" });
    const ex1Search = await db.search({
      term: "quick",
      properties: ["example"],
    });

    expect(ex1Insert.id).toBeDefined();
    expect(ex1Search.count).toBe(1);
    expect(ex1Search.elapsed).toBeDefined();
    expect((ex1Search.hits[0] as any).example).toBe("The quick, brown, fox");
  });

  it("should correctly paginate results", async () => {
    const db = new Lyra({
      schema: {
        animal: "string",
      },
    });

    await db.insert({ animal: "Quick brown fox" });
    await db.insert({ animal: "Lazy dog" });
    await db.insert({ animal: "Jumping penguin" });
    await db.insert({ animal: "Fast chicken" });
    await db.insert({ animal: "Fabolous ducks" });
    await db.insert({ animal: "Fantastic horse" });

    const search1 = await db.search({ term: "f", limit: 1, offset: 0 });
    const search2 = await db.search({ term: "f", limit: 1, offset: 1 });
    const search3 = await db.search({ term: "f", limit: 1, offset: 2 });
    const search4 = await db.search({ term: "f", limit: 2, offset: 2 });

    expect(search1.count).toBe(4);
    expect((search1.hits[0] as any).animal).toBe("Quick brown fox");

    expect(search2.count).toBe(4);
    expect((search2.hits[0] as any).animal).toBe("Fast chicken");

    expect(search3.count).toBe(4);
    expect((search3.hits[0] as any).animal).toBe("Fabolous ducks");

    expect(search4.count).toBe(4);
    expect((search4.hits[0] as any).animal).toBe("Fabolous ducks");
    expect((search4.hits[1] as any).animal).toBe("Fantastic horse");
  });

  it("Should throw an error when searching in non-existing indices", async () => {
    const db = new Lyra({ schema: { foo: "string", baz: "string" } });

    try {
      await db.search({
        term: "foo",
        properties: ["bar"],
      });
    } catch (err) {
      expect(err).toMatchInlineSnapshot(
        `"Invalid property name. Expected a wildcard string (\\"*\\") or array containing one of the following properties: foo, baz, but got: bar"`
      );
    }
  });

  it("Should correctly remove a document after its insertion", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
      },
    });

    const { id: id1 } = await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    const { id: id2 } = await db.insert({
      quote:
        "To live is the rarest thing in the world. Most people exist, that is all.",
      author: "Oscar Wilde",
    });

    await db.insert({
      quote: "So many books, so little time.",
      author: "Frank Zappa",
    });

    const res = await db.delete(id1);

    const searchResult = await db.search({
      term: "Oscar",
      properties: ["author"],
    });

    expect(res).toBeTruthy();
    expect(searchResult.count).toBe(1);
    expect((searchResult.hits[0] as any).author).toBe("Oscar Wilde");
    expect((searchResult.hits[0] as any).quote).toBe(
      "To live is the rarest thing in the world. Most people exist, that is all."
    );
    expect((searchResult.hits[0] as any).id).toBe(id2);
  });

  it("Should preserve identical docs after deletion", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
      },
    });

    const { id: id1 } = await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    const { id: id2 } = await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    await db.insert({
      quote: "So many books, so little time.",
      author: "Frank Zappa",
    });

    const res = await db.delete(id1);

    const searchResult = await db.search({
      term: "Oscar",
      properties: ["author"],
    });

    const searchResult2 = await db.search({
      term: "already",
      properties: ["quote"],
    });

    expect(res).toBeTruthy();
    expect(searchResult.count).toBe(1);
    expect((searchResult.hits[0] as any).author).toBe("Oscar Wilde");
    expect((searchResult.hits[0] as any).quote).toBe(
      "Be yourself; everyone else is already taken."
    );
    expect((searchResult.hits[0] as any).id).toBe(id2);

    expect(searchResult2.count).toBe(1);
    expect((searchResult2.hits[0] as any).author).toBe("Oscar Wilde");
    expect((searchResult2.hits[0] as any).quote).toBe(
      "Be yourself; everyone else is already taken."
    );
    expect((searchResult2.hits[0] as any).id).toBe(id2);
  });

  it("Should be able to insert documens with non-searchable fieldsa", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
        isFavorite: "boolean",
        rating: "number",
      },
    });

    await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
      isFavorite: false,
      rating: 4,
    });

    await db.insert({
      quote: "So many books, so little time.",
      author: "Frank Zappa",
      isFavorite: true,
      rating: 5,
    });

    const search = await db.search({
      term: "frank",
    });

    expect(search.count).toBe(1);
    expect((search.hits[0] as any).author).toBe("Frank Zappa");
  });

  it("Should exact match", async () => {
    const db = new Lyra({
      schema: {
        author: "string",
        quote: "string",
      },
    });

    await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    const partialSearch = await db.search({
      term: "alr",
      exact: true,
    });

    expect(partialSearch.count).toBe(0);

    const exactSearch = await db.search({
      term: "already",
      exact: true,
    });

    expect(exactSearch.count).toBe(1);
    expect((exactSearch.hits[0] as any).quote).toBe(
      "Be yourself; everyone else is already taken."
    );
    expect((exactSearch.hits[0] as any).author).toBe("Oscar Wilde");
  });
});
