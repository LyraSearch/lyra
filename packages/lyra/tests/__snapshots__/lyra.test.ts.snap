// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkInsertDocSchema should compare the inserted doc with the schema definition 1`] = `
"Invalid document structure. 
Class has been initialized with the following schema: 

{
  \\"quote\\": \\"string\\",
  \\"author\\": \\"string\\"
}

but found the following doc:

{
  \\"quote\\": \\"hello, world!\\",
  \\"author\\": true
}"
`;

exports[`checkInsertDocSchema should compare the inserted doc with the schema definition 2`] = `
"Invalid document structure. 
Class has been initialized with the following schema: 

{
  \\"quote\\": \\"string\\",
  \\"author\\": \\"string\\"
}

but found the following doc:

{
  \\"quote\\": \\"hello, world!\\",
  \\"authors\\": \\"author should be singular\\"
}"
`;

exports[`checkInsertDocSchema should compare the inserted doc with the schema definition 3`] = `
"Invalid document structure. 
Class has been initialized with the following schema: 

{
  \\"quote\\": \\"string\\",
  \\"author\\": \\"string\\"
}

but found the following doc:

{
  \\"quote\\": \\"hello, world!\\",
  \\"foo\\": {
    \\"bar\\": 10
  }
}"
`;

exports[`defaultLanguage should throw an error if the desired language is not supported 1`] = `
"Language \\"latin\\" is not supported.
Supported languages are:
 - dutch
 - english
 - french
 - italian
 - norwegian
 - portugese
 - russian
 - spanish
 - swedish"
`;
